### ISSUE TITLE
Unity 5 - RuntimeInitializeOnLoadMethodAttribute

### ISSUE THREAD

**Issue by rotorz** [ssannandeji/Zenject-2019#31]
Hey there

Just a heads up, Unity 5 has added a new attribute `RuntimeInitializeOnLoadMethodAttribute` which allows you to have a main entry point into the game.

Possibly a good place to initialize the global composition root in Unity 5?

**Comment by svermeulen** [comment]
Yeah this could work nicely.  That might allow for Zenject to properly inject all game objects in the scene before the Awake event occurs (currently there's no guarantee of this).  Thanks for the heads up.

**Comment by rotorz** [comment]
I have been experimenting with this new `RuntimeInitializeOnLoadMethodAttribute` and due to what I can only believe is a bug in its implementation have found that it is somewhat useless since it is executed AFTER all the objects in the scene have awoken with `Awake()`.

I have reported this as a bug with the Unity bug tracker.

Another developer has reported this via the Unity feedback website:
http://feedback.unity3d.com/suggestions/runtimeinitializeonloadmethod-called-too-late

I will post back here if they decide to fix this :)

**Comment by Zexbe** [comment]
It has been added to Unity
http://docs.unity3d.com/ScriptReference/RuntimeInitializeOnLoadMethodAttribute-loadType.html
http://docs.unity3d.com/ScriptReference/RuntimeInitializeLoadType.html

**Comment by Zexbe** [comment]
@svermeulen What I mean is, the "Before" Awake implementation has been added.

**Comment by extrawurst** [comment]
here is the actual doc for how to use that: http://docs.unity3d.com/ScriptReference/RuntimeInitializeOnLoadMethodAttribute-ctor.html

looks like a great way to relieably inject and still allow awake to be used

**Comment by svermeulen** [comment]
I looked into this some more.  I confirmed that it does get triggered before Awake, which is good, but this only happens once on startup.  When you change scenes, it doesn't get called.  So if we were considering using this to initialize the SceneCompositionRoot, we'd have to figure out some other way to hook into scene changes, so we can initialize before Awake in those cases as well.  I also looked at the `OnLevelWasLoaded` but that seems to get called after awake.  

In other words, there's a main entry point before Awake but there doesn't appear to be a scene entry point before Awake.

BUT we can at least use this for the global composition root so I'll take a look at that.

**Comment by extrawurst** [comment]
maybe we have some way in unity 5.4: 

https://unity3d.com/unity/beta
"`Scene Management: Added events onSceneLoaded, onSceneUnloaded, onActiveSceneChanged to SceneManager`"

**Comment by svermeulen** [comment]
That's a good thought but apparently `onActiveSceneChanged` and `onSceneLoaded` both also occur after Awake (tested it just now)

**Comment by svermeulen** [comment]
Ok I've changed `GlobalCompositionRoot` to use `[RuntimeInitializeOnLoadMethod]`.  Thanks for the suggestions on this, it seems to work well.  One interesting result of this is that it will be run regardless of whether you are using `SceneCompositionRoot` in any of your scenes.

@Extrawurst I also wanted to mention that as long as you have your script execution order set properly, you will be guaranteed to have Zenject run its Awake first, so you can have MonoBehaviours that do stuff in Awake and also rely on Zenject.   The script execution order should be set up to do this by default.

But it would still be nice if we could somehow execute scene installers before awake - that would open up some interesting design choices.  I'll keep an eye out for new Unity releases to see if they do add something like that.  They seemed to almost add that with `OnLevelWasLoaded` so maybe they will at some point.
